src/mono/System.Private.CoreLib/src/System/AppDomain.Mono.cs

```diff
@@ -0,0 +1,15 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System
+{
+    public partial class AppDomain
+    {
+#if TARGET_ANDROID
+        internal void DoUnhandledException (UnhandledExceptionEventArgs args)
+        {
+            AppContext.DoUnhandledException (this, args);
+        }
+#endif
```

Comment:
Where is this called and why is it Android specific?

Reply:
I don't fully understand how this stuff works, but xam-android is catching all exceptions in https://github.com/xamarin/xamarin-android/blob/main/src/Mono.Android/Android.Runtime/JNINativeWrapper.cs#L52. This wrapper might call into `Mono_UnhandledException` for the debugger stuff and later it calls into `AndroidEnvironment.UnhandledException`. Sometime later, not sure how, it will end up in https://github.com/xamarin/xamarin-android/blob/main/src/Mono.Android/Android.Runtime/JNIEnv.cs#L297

Reply:
I don't see that code used anywhere and I don't think it should be Android-specific. Handling of unhandled exceptions should work everywhere.

Reply:
@marek-safar [wrote](https://github.com/dotnet/runtime/pull/55904/files#r672272483):

> I don't see that code used anywhere

I don't know which code you're referring to specifically, given that @BrzVlad linked to relevant places of code.  Thus, an example: suppose you have the following Xamarin.Android code:

```csharp
using Android.App;
using Android.OS.;
using System;

[Activity (MainLauncher = true)
partial class MyActivity : Activity {
    public override void OnCreate (Bundle savedInstanceState) => throw new System.Exception ("Yolo!");
}
```

At runtime, the Java Callable Wrapper for `MyActivity` will be created, which will cause the C# `MyActivity` constructor to be executed.  Later, Android will invoke `Activity.onCreate()` on that instance, which will invoke `MyActivity.OnCreate()` to be invoked.

On entry to `MyActivity.OnCreate()`, we have a stack frame of:

 1. Java [caller of `Activity.onCreate()`
 2. `JNINativeWrapper`-generated delegate instance
 3. A binding generated "marshal method", `Activity.n_OnCreate_Landroid_os_Bundle_()`
 4. `MyActivity.OnCreate()`

(2) is effectively:

```csharp
static void n_OnCreate(IntPtr env, IntPtr self, IntPtr bundle)
{
    try {
        Activity.n_OnCreate_Landroid_os_Bundle_ (env, self, bundle);
    }
    catch (Exception e) {
        AndroidEnvironment.UnhandledException (e);
    }
}
```

(3) would be:

```csharp
partial class Activity {
	static void n_OnCreate_Landroid_os_Bundle_ (IntPtr jnienv, IntPtr native__this, IntPtr native_savedInstanceState)
	{
		var __this = global::Java.Lang.Object.GetObject<Android.App.Activity> (jnienv, native__this, JniHandleOwnership.DoNotTransfer)!;
		var savedInstanceState = global::Java.Lang.Object.GetObject<Android.OS.Bundle> (native_savedInstanceState, JniHandleOwnership.DoNotTransfer);
		__this.OnCreate (savedInstanceState);
	}
}
```

[`AndroidEnvironment.UnhandledException(e)`](https://github.com/xamarin/xamarin-android/blob/1b71da0cd3a44861b006ac41c77a24eedf066281/src/Mono.Android/Android.Runtime/AndroidEnvironment.cs#L100-L120) will basically do:

```csharp
JNIEnv.Throw(new JavaProxyThrowable(e));
```

When `MyActivity.OnCreate()` throws the exception, we'll run the `catch` block in (2).  `JNIEnv.Throw()` causes there to be a "pending exception"; on return to Java, Java will raise the exception, allowing Java-side `catch` and `finally` blocks to participate in the stack unwind.

This will cause frame (1) and callers of (1) to eventually unwind, potentially unwinding the entire Java-side stack.  When that happens, Java will eventually execute:

```java
java.lang.Thread.getUncaughtExceptionHandler()
    .uncaughtException(currentThread, currentException);
```

During startup, Xamarin.Android registers an uncaught exception handler with Java -- the `UncaughtExceptionHandler` type which @BrzVlad linked to -- and `UncaughtExceptionHandler.UncaughtException()` will raise the `AppDomain.UnhandledException` event, by calling `AppDomain.DoUnhandledException(UnhandledExceptionEventArgs)`.

We *don't* want the `JNINativeWrapper`-emitted delegate to raise `AppDomain.UnhandledException`, because it might not be unhandled!  If we're in a Java > Managed > Java > Managed stack, one of the calling stack frames might catch the exception.

We thus need *a* way for Xamarin.Android to cause the `AppDomain.UnhandledException` event to be raised.

Related? https://github.com/xamarin/xamarin-android/pull/4877/files?short_path=87e7d90#diff-87e7d904d49ad72bcc40356d159cf46a7508f28adf19a8cf2707d109f223adb7

Reply:
Apparently the implicit question was "how do we actually register the Java-side unhandled exception support"?

The `mono.android.Runtime` class is used *everywhere* in a Xamarin.Android app, including as part of process bootstrap.  The [`Runtime` static constructor](https://github.com/xamarin/xamarin-android/blob/1b71da0cd3a44861b006ac41c77a24eedf066281/src/java-runtime/java/mono/android/Runtime.java#L14) registers a [`Thread.UncaughtExceptionHandler` implementation](https://github.com/xamarin/xamarin-android/blob/1b71da0cd3a44861b006ac41c77a24eedf066281/src/java-runtime/java/mono/android/Runtime.java#L28-L44), and implements `Thread.UncaughtExceptionHandler.uncaughtException()` by [calling `Runtime.propagateUncaughtException()`](https://github.com/xamarin/xamarin-android/blob/1b71da0cd3a44861b006ac41c77a24eedf066281/src/java-runtime/java/mono/android/Runtime.java#L39).

`Runtime.propagateUncaughtException()` calls [`JNIEnv.PropagateUncaughtException()`](https://github.com/xamarin/xamarin-android/blob/main/src/Mono.Android/Android.Runtime/JNIEnv.cs#L270-L301), which calls `AppDomain.DoUnhandledException()`.

Reply:
As per [above conversation](https://github.com/dotnet/runtime/pull/55904/files#r672546220), I *think* we don't actually need `AppDomain.DoUnhandledException()`; xamarin-android could instead (presumably) be updated to call `mono_unhandled_exception()` instead.

Reply:
@grendello will work on a change to xamarin-android wherein we call `mono_unhandled_exception()` instead of `AppDomain.DoUnhandledException()`: https://discord.com/channels/732297728826277939/732297837953679412/866765321821618256

Reply:
Eventually, does it make sense to do something similar to XI?

Register: 

https://github.com/xamarin/xamarin-macios/blob/2bee92225cf1ec3695cfdd9e60618a5f9bb7dfdf/src/ObjCRuntime/Runtime.CoreCLR.cs#L102

And this is the runtime API it's using 

https://github.com/dotnet/runtime/blob/e0024e22c7761cbbdd10a845f8b0269ea654cbfc/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/ObjectiveC/ObjectiveCMarshal.cs#L60

Reply:
Let's try to use the existing embedding API first before considering this PR.